lib/streamlit/testing/v1/app_test.py

```diff
@@ -0,0 +1,266 @@
+# limitations under the License.
+from __future__ import annotations
+
+import hashlib
+import pathlib
+import tempfile
+import textwrap
+import traceback
+from typing import Any, Sequence
+from unittest.mock import MagicMock
+
+from streamlit import source_util
+from streamlit.proto.WidgetStates_pb2 import WidgetStates
+from streamlit.runtime import Runtime
+from streamlit.runtime.caching.storage.dummy_cache_storage import (
+    MemoryCacheStorageManager,
+)
+from streamlit.runtime.media_file_manager import MediaFileManager
+from streamlit.runtime.memory_media_file_storage import MemoryMediaFileStorage
+from streamlit.runtime.state.session_state import SessionState
+from streamlit.testing.v1.element_tree import (
+    Block,
+    Button,
+    Caption,
+    Checkbox,
+    Code,
+    ColorPicker,
+    DateInput,
+    Divider,
+    ElementList,
+    ElementTree,
+    Exception,
+    Header,
+    Latex,
+    Markdown,
+    Multiselect,
+    Node,
+    NumberInput,
+    Radio,
+    Selectbox,
+    SelectSlider,
+    Slider,
+    Subheader,
+    Text,
+    TextArea,
+    TextInput,
+    TimeInput,
+    Title,
+    WidgetList,
+)
+from streamlit.testing.v1.local_script_runner import LocalScriptRunner
+from streamlit.testing.v1.util import patch_config_options
+
+TMP_DIR = tempfile.TemporaryDirectory()
+
+
+class AppTest:
+    def __init__(self, script_path: str, default_timeout: float):
+        self._script_path = script_path
+        self.default_timeout = default_timeout
+        self.session_state = SessionState()
+        self.query_params: dict[str, Any] = {}
+
+        tree = ElementTree()
+        tree._runner = self
+        self._tree = tree
+
+    @classmethod
+    def from_string(cls, script: str, default_timeout: float = 3) -> AppTest:
+        """Create a runner for a script with the contents from a string.
+
+        Useful for testing short scripts that fit comfortably as an inline
+        string in the test itself, without having to create a separate file
+        for it.
+
+        `default_timeout` is the default time in seconds before a script is
+        timed out, if not overridden for an individual `.run()` call.
+        """
+        hasher = hashlib.md5(bytes(script, "utf-8"))
+        script_name = hasher.hexdigest()
+
+        path = pathlib.Path(TMP_DIR.name, script_name)
+        aligned_script = textwrap.dedent(script)
+        path.write_text(aligned_script)
+        return AppTest(str(path), default_timeout=default_timeout)
+
+    @classmethod
+    def from_file(cls, script_path: str, default_timeout: float = 3) -> AppTest:
+        """Create a runner for the script with the given name, for testing.
+
+        `default_timeout` is the default time in seconds before a script is
+        timed out, if not overridden for an individual `.run()` call.
+        """
+        # TODO: Make this not super fragile
+        stack = traceback.StackSummary.extract(traceback.walk_stack(None))
```

Comment:
nit: What path is this trying to resolve to? Maybe add a comment here?

Reply:
It is trying to find the file the test is in, so we can make file path strings resolve relative to there. But this stack track inspection is so potentially fragile that I'm thinking about giving up on that QoL feature just so it won't randomly break if someone tries to do things slightly differently than I do